python/ray/experimental/channel/nccl_group.py

```diff
@@ -188,22 +253,40 @@ def recv(
             raise RayChannelError("NCCL group has been destroyed.")
         assert allocator is not None, "NCCL group requires a tensor allocator"
         buf = allocator(shape, dtype)
-        self._comm.recv(
-            self.nccl_util.get_tensor_ptr(buf),
-            buf.numel(),
-            self.nccl_util.get_nccl_tensor_dtype(buf),
-            peer_rank,
-            self._cuda_stream.ptr,
-        )
 
-        # Buffer values are undefined if NCCL ops are aborted. Therefore, we
-        # need to synchronize here and check that the channel is still open to
-        # ensure that the receive buffer is valid.
-        # TODO(swang): Avoid CUDA synchronization.
-        self._cuda_stream.synchronize()
-        if self._closed:
-            raise RayChannelError("NCCL group has been destroyed.")
-        return buf
+        if self._use_communication_streams:
```

Comment:
Nit: Can you try to collapse these branches so that they use more or less the same logic? The different branches make it easier to introduce bugs.

I think the main differences are:
- whether we synchronize the stream after the op or not
- whether we create an event or not (btw if you follow the earlier suggestion about wrapping this into GPUFuture constructor, then we don't need to do that here at all)

Reply:
Done.